### DISCUSSION TITLE
HTMLy Comments System

### DISCUSSION THREAD

**Discussion by KuJoe** [danpros/htmly#803] [Ideas]
There's been multiple issues open requesting alternatives to Disqus and Facebook for integrated comments so I wanted to get a discussion going for this. I explored a few of the recommended solutions and even built the integration for one, but I feel like this needs more input from users before deciding on a solution.

1. [CGrassin/interact_php](https://github.com/CGrassin/interact_php) - Very basic with minimal spam filtering and moderation. Last updated >1 year ago. 1 outstanding issue (feature) and 0 outstanding PRs.
2. [djyde/cusdis](https://github.com/djyde/cusdis) -  Could not get the Docker container build or the source build working. Integration with the SaaS offering was simple, but looked out of place with the default CSS styling. Last updated >1 year ago. 83 outstanding issues (including multiple install errors) and 23 outstanding PRs.
3. [isso-comments/isso](https://github.com/isso-comments/isso) - Appears to be the most feature rich with multiple install options (Docker, pip, and source). Last updated 4 months ago. 52 outstanding issues (lots of feature requests) and 9 outstanding PRs.
4. [GiovanniSalmeri/yellow-comment](https://github.com/GiovanniSalmeri/yellow-comment) - Appears to be an extension for [datenstrom/yellow](https://github.com/datenstrom/yellow) but looks very lightweight, unsure how hard it would be to integrate into HTMLy. Last updated 4 months ago. 1 outstanding issue (feature) and 0 outstanding PRs.

**Comment by KuJoe** [comment]
Would a hooks system be a good idea? This way we can have higher compatibility with current and future themes while supporting more options for users without having to integrate them all. We can even remove the existing Disqus and Facebook integration and replace them with separate files.

Effectively we would replace this code in **post.html.php**:
```
<?php if (disqus()): ?>
    <?php echo disqus($p->title, $p->url) ?>
<?php endif; ?>
<?php if (disqus_count()): ?>
    <?php echo disqus_count() ?>
<?php endif; ?>
<?php if (facebook() || disqus()): ?>
<div class="comments-area" id="comments">
    <?php if (facebook()): ?>
        <div class="fb-comments" data-href="<?php echo $p->url ?>" data-numposts="<?php echo config('fb.num') ?>" data-colorscheme="<?php echo config('fb.color') ?>"></div>
    <?php endif; ?>
    <?php if (disqus()): ?>
        <div id="disqus_thread"></div>
    <?php endif; ?>
</div>
<?php endif; ?>
```

With this code:
```
<?php post_hooks($p); ?>
```

And adding this function to **functions.php**:
```
function post_hooks($params = null)  
{
    $directory = __DIR__ . '/hooks/posts/';
    $files = scandir($directory);
    
    foreach ($files as $file) {
        if (pathinfo($file, PATHINFO_EXTENSION) === 'php') {
            include($directory . '/' . $file);
        }
    }
    if (!$file) {
        return false;
    }
}
```

So any files within **system\includes\hooks\posts** will be included into the page (this can be done for all of the other pages and different parts of each also).

Here's an example of a file called **comments.php**:
```
<?php

echo '<div class="comments-area" id="comments"><div class="fb-comments"><div id="cusdis_thread"
  data-host="https://cusdis.com"
  data-app-id="<MY_CUSDIS_APPID>"
  data-page-id="'.$params->slug.'"
  data-page-url="'.$params->url.'"
  data-page-title="'.$params->title.'"
></div>
<script async defer src="https://cusdis.com/js/cusdis.es.js"></script></div></div>';

?>
```

The end result:
![20240908031135_JDYOGA7I](https://github.com/user-attachments/assets/6f0ebd97-b0f9-4a4f-b1a6-d3692f641f33)

Since it uses any files in that directory, here's an example of a "Hello World" script I uploaded to the same directory, this adds some versatility without requiring direct integration into HTMLy and without having to manually edit theme files with custom changes every time they get updated.
![image](https://github.com/user-attachments/assets/13ccc2e8-04f4-4659-a044-183a857e4498)

**Comment by Joduai** [comment]
yesterday I had a search session for self-hosted php flat file comments system.  
Years ago there were plenty of them, but as technology advances, and changes php isn't such popular choice as it was 10yrs ago.  
Despite already mentioned, there are also Jskomment, and mentioned in issues hashover.  

Isso is a python solution, thus I wouldn't even consider this for a non techie blog user.  
Comments plugin for yellow cms seems to be very simple. It's obviously one of its pros. But having a glimpse at its source code there's only form validation as a way of filtering spam. I have no idea if yellowcms has builtin solutions like captcha and this plugin utilises it, as I touched this cms once circa 5yrs ago.  

Cusdis is a nodejs based system. No chances for this on typical LAMP shared hosting providers.

After wasting so much time I found out somewhat a complete comment system list with all scripts I've already found and looked into:  
https://lisakov.com/projects/open-source-comments/
  
I even thought about looking at guestbooks scripts to adapt it as a comment system, but such "widgets" are relicts of the past.  
  
Basically we can choose from interact, and hashover(next, as it's v.2). 
Every of those lack PR approving, development is some kind of stale. I'm afraid of this as things like captchas that base on 3rd party solutions like recaptcha, akismet, hcatpcha, mosparo etc. need support when their api changes. I didn't dig yet, but builtin simple but smart enough QA solution would be good enough to prevent automated spam attempts.  
Then goes the support for newer php versions...

integrating comments menagement into admin panel is next question. Having to manage every addon on its own is a managment killer. I was wondering about adding reactions to posts. One of comments system had this builting.Can't remind which one. 
I've implemented external uploader and I load it through ajax iframe within admin panel, but this isn't a solution to manage every external script.


As for the hooks you are proposing, isn't it more a part of supporting plugins?  
This would contradict basic htmly principle with no plugin support.  
Nonetheless scanning a directory and automatically launch every scirpt inside this is a bit dangerous.  
I've seen such solutions for plugin support so that every scanned script had to be enabled firstly to be executed in cms.

**Comment by thopanX** [comment]
I installed Interact, but have the problem that comments aren't post-specific. The same test comments are shown on every post.

The comment system wants an identfier, what's the code to include the post id?

**Reply by Joduai** [reply]
If you don't specify ID when calling `\Interact_PHP\Interact_PHP();` comment's file name is generated from page's URI without domain name, not recommended thou. 
1. Do you call the code in post.html.php file?
2. Do you call interact with empty ID for default behaviour?

The simplest id is page slug, or full url if you want the comments to be available also on static page.
Try with `$post->slug` or` $post->url` as ID

**Reply by thopanX** [reply]
1. yes
2. yes

The ID for post slugs works perfectly, thanks!

**Comment by thopanX** [comment]
Has anybody gotten Interact's admin interface to accept a password via hash in the settings.php? I triple-checked and my hash seems to be alright, but I keep getting the info that the password is wrong.

**Reply by Joduai** [reply]
try this hash generator for a temporary password
https://emn178.github.io/online-tools/sha256.html

**Reply by thopanX** [reply]
Thanks, that worked! No idea why the hash created on the CLI here didn't. Anyway, great help!

**Comment by Clover-Zero** [comment]
This is a bit different from the other comment systems and I'm probably the only one who wants it, but, I'm considering to add [Mastodon comments](https://danielpecos.com/2022/12/25/mastodon-as-comment-system-for-your-static-blog/) ([another guide](https://bechster.dk/mastodon-kommentarer-i-publii/)). It seems pretty easy to apply to static sites, but I'm afraid I'm not savvy enough to know how to apply it to HTMLy.

**Reply by thopanX** [reply]
No, you're definitely not alone in wanting to use Fediverse comments!

**Reply by Clover-Zero** [reply]
Glad to hear that! I hope someone can help with that. ( ; ω ; )

**Reply by Clover-Zero** [reply]
I got it working!! Turns out it's pretty easy and simple lol:

1. Add Mastodon-comments as the guide said to `<head>`.
2. Make a custom field (in this case, `toot_id`).
3. Add `<mastodon-comments host="instance" user="user" tootId="<?php echo get_field('toot_id', $p->raw);?>"></mastodon-comments>` to `post.html.php`.
4. Fill in `toot_id` custom field in your blog post with your toot's ID. Done!

**Reply by danpros** [reply]
That's great! Custom fields open up a lot of possibilities 😃

**Comment by rolandixor** [comment]
Are there any of these suggested solutions that:

1. Would work for users without an account
2. Would be possible to integrate into a custom theme, today?

I'd like to try out one of these solutions on my blog (and in custom themes).

**Comment by mithat** [comment]
I want to throw out another potential alternative. [Flatpress](https://github.com/flatpressblog/flatpress) (also GPL-2) has a baked-in [comments system](https://github.com/flatpressblog/flatpress/blob/master/comments.php) that when I used Flatpress was usable (though the anti-spam might be too light for some users). Assuming porting it to HTMLy is workable, is this a solution worth considering?

**Comment by afoster1955** [comment]
I have had many problems getting some of the third party comment systems to work with HTMLy but recently I stumbled across Hashover (2.0) and it worked right away.  I read in this thread that Hashover is not an optimal system, but it works for me and very easy to integrate into the blog for a non techie.

**Reply by mithat** [reply]
The most recent update to the [1.0 Github repo](https://github.com/jacobwb/hashover) was 6 years ago, and to the [2.0 Github repo](https://github.com/jacobwb/hashover-next) was 3 years ago. In practice this means both might have unpatched security issues and/or a PHP update that removes deprecated features might break the system.

It should also be noted that Hashover is licensed as GNU Affero, which comes with it some obligations that the plain GPLs don't.

**Reply by afoster1955** [reply]
Good to know, thanks.  I wish that HTMLy had its own comment system so that there is no need for any third party involvement. Hashover is especially good (for me anyway) because there is little to no coding needed to get it to work.

**Comment by andrigamerita** [comment]
I'm new to HTMLy, but one of the first things I did on my website has been [setting up comments my own way](https://stuff.octt.eu.org/2025/06/integrating-giscus-comments-into-my-htmly-website), because I too think that Disqus (not to mention Facebook) nowadays are really not ideal.

I'd be interested to see if some proper generic integration can be implemented (especially to allow anonymous comments), but for my own use case right now I'm happy with [Giscus](https://giscus.app/), which just uses GitHub Discussions and has a public instance, so doesn't require to install anything on the website server. I'm surprised no one has ever mentioned it yet in this discussion, given we're all GitHub users here!

**Comment by Joduai** [comment]
You might check https://github.com/splitbrain/meh/  
A JS+PHP>8.2+SQLite comment system.   
I had problems compiling it under freebsd, thus asked dev for a ready to deploy package, so check releases if you aren't familiar or do not have access to composer, and npm package manager, or get compilation errors due to compatibility.   
Although shell access is a must have to initialize the sqlite file database, and some basic configuration.   
  
Unfortunately I did not succeed to make it work due to json parse data error, which I got no idea how to debug as this comment system communicates through avascript(?) api. It still might be a compat problem with freebsd, and I don't have enough time to put up even a local test environment to find the culprit. There are some security features like time blockade allowing to post after 30sec by default.  

wish you luck

**Reply by afoster1955** [reply]
At the moment, I am trying to get the interact_php comment system working.  I have it being displayed at the bottom of the article but it is inconsistent in getting any comments posted.  Sometimes it does and other times it does not. Also, I am not certain I have the correct parameters to enter into the comment system to make it work separately with each article rather than displaying the same comment after all articles.  I really like HTMLy, but not having a comment system is making me look elsewhere.  I realize it works with Disquus, Facebook and Commentics, but all of those are bloated and not good for my readers.  I will keep trying, but I'm fast running out of patience.  Your suggested solution is not viable for me as my host does not (at the moment) provide access to php 8.x.x

**Comment by Emidio** [comment]
Adding my 2 cents to this thread. I'm switching to HTMLy, and looking for a very simple comment system too to be integrated. I don't like external systems (European GDPR and privacy concerns) and what I found as flat file PHP comment systems is mainly abandonware. I wrote a very simple threaded flat file commenting system integrated with HTMLy (slightly modified 3.1.1 code) and a template (derived from Occasio). 

Still a work in progress (some CSS to be tweaked/fixed, security hardening, notifications) but already a decent alternative. Not sure someone can be interested, and could make sense to propose HTMLy integration (that I'd like, so I don't have to modify code after each upgrade).

Attached some screenshots.

<img width="2929" height="2324" alt="comments-admin-all" src="https://github.com/user-attachments/assets/49e8effb-3bba-4856-a70a-6735ecc524a0" />
<img width="2929" height="2378" alt="comments-admin-config" src="https://github.com/user-attachments/assets/129f79c4-8c15-4296-a1ac-dc48c85aab75" />
<img width="2929" height="2324" alt="comments-admin-pending" src="https://github.com/user-attachments/assets/622f42ec-9088-4447-adc4-0d4a946c3891" />
<img width="2929" height="5318" alt="comments-public" src="https://github.com/user-attachments/assets/6b5afec7-6666-453c-8728-93dd619c1671" />

**Reply by Joduai** [reply]
>Not sure someone can be interested

someone?
everyone in this thread will be interested! ;)

I'm just about getting to update my old htmly instance with couple tweaks so I'd gladly try out if you provide a downloadable package.  
Dare to list features you implemented and working on?  
Unfortunately securing forms allowing different types of captcha is somewhat a must have nowadays.

**Comment by thopanX** [comment]
That looks awesome. Will try it out these days.
 
"Emidio" ***@***.*** – 7. November 2025 10:43 

>  
> Adding my 2 cents to this thread. I'm switching to HTMLy, and looking
> for a very simple comment system too to be integrated. I don't like
> external systems (European GDPR and privacy concerns) and what I found
> as flat file PHP comment systems is mainly abandonware. I wrote a very
> simple threaded flat file commenting system integrated with HTMLy
> (slightly modified 3.1.1 code) and a template (derived from Occasio).
> Still a work in progress (some CSS to be tweaked/fixed, security
> hardening, notifications) but already a decent alternative. Not sure
> someone can be interested, and could make sense to propose HTMLy
> integration (that I'd like, so I don't have to modify code after each
> upgrade).
> Attached some screenshots.
> comments-admin-all.png (view on web) comments-admin-config.png (view on
> web) comments-admin-pending.png (view on web) comments-public.png (view
> on web)
> —Reply to this email directly, view it on GitHub, or unsubscribe.You
are
> receiving this because you commented.Message ID:
> ***@***.***>
>

**Comment by danpros** [comment]
@Emidio That's awesome! Please create a pull request so we can test it 😃

**Reply by Emidio** [reply]
Hi Dan - during the weekend I'll update the fork I just did, then I'll create the pull request. I'm adding the modifications to HTMLy and the Ovi theme I created to handle comments.

**Reply by danpros** [reply]
Thanks. I look forward for it.

**Reply by Emidio** [reply]
Done. I'm dubious about PHPMailer/mail notifications to admin. Have a look.

**Reply by danpros** [reply]
Thanks. I already merge it so anyone can debug it. For PHP mailer its fine, we can use it eg. for contact form widget in the future. But please use the composer instead.

**Reply by Emidio** [reply]
Hi Dan. It was installed using composer. I had a quick look, I can see it in composer.json. Just one note: PHPMailer is not latest version due to PHP compatibility 7.2 (error raised by composer at first install).

**Reply by danpros** [reply]
Hi, yes apologies. I didn't check it carefully

**Comment by rolandixor** [comment]
Would these comments work with other themes, provided we do the necessary
CSS modifications? Meaning, is it a simple include that we can drop into
any theme and work from there?

https://rolandixor.pro/portfolio
https://rolandixor.pro

**Reply by Emidio** [reply]
Just a few lines of PHP to be added to post.html.php.

    <?php if (local()): ?>
        <?php
        $postId = str_replace(site_url(), '', $p->url);
        displayCommentsSection($postId);
        ?>
    <?php endif; ?>

Should be some more after I finish writing success/fail messages after commenting. The main work will be CSS to match it to your theme. And consider some CSS cleaning, I manly work with databases/API/backends, CSS is a necessary annoyance to me :-)

**Reply by rolandixor** [reply]
> CSS is a necessary annoyance to me :-)

CSS is my love language haha. I can probably pitch in on some themes here and here if this gets implemented to core. Definitely would want to get it working with my own theme (which I haven't yet released)

**Reply by Emidio** [reply]
@danpros Hi Dan. I updated my fork with several fixes and improvements on the comments system. Just tell if you like a pull request or wait some time. New release use a different comment naming system (to better match .md files HTMLy file names) and handle admin comments list pagination and .md files renaming/delete support (the associated comment file is renamed accordingly). Also, to display comments in a blog post, now use:

    <?php if (local() && comments_config('comments.show.posts') === 'true'): ?>
        <?php
        displayCommentsSection($p->url, $p->file);
        ?>
    <?php endif; ?>

**Reply by danpros** [reply]
@Emidio Yes, please create a PR request for that.

It's recommended to switch to using only one function when calling comments in the future. The old function is still available for backward compatibility.

Example:

```php
<?php if (comments()):?>
    <?php echo comments($p);?>
<?php endif;?>
```

Inside the `comments()` function, we can specify which comments to display based on the selected configuration.

In `$p` we can extract the value. We can even tell whether it's on a blog post, page or subpage. If blog post the `$p->author` is available. If page the `$p->parent` is `null` etc. 

With this we can separate if this is a comment for a blog post, page and subpage (for example by adding `post_thePostSlug` as the file name, `page_thePageSlug`, `subpage_theSubPageSlug` etc. or separate folder). Example: `content/comments/post` folder.

And to get the slug for blog post to display it to the admin, example:

```php
$post = find_post(null, null, $comment['post_id']); 
$current_post = $post['current'];
```

Become:

```php
<td>
    <a href="<?php echo $current_post->url;?>" target="_blank">
        <?php echo _h($comment['post_id']); ?>
    </a>
</td>
```
 
Edit: so for the post_id us the `$p->slug;`

**Reply by Emidio** [reply]
Hi Dan. Good points, I perfectly agree. Now I almost finished development, and I'll concentrate on cleaning and better integration. As for $comment['post_id'], I actually changed the functions to work both with actual url or .md file path (both available, second option is faster and more direct). I'll keep the comments json file naming I actually implemented in my current version (changed from last PR), but surely I'd like to make it more "standard" and integrated in HTMLy - I just don't know it enough to get the correct calls sometimes, probably I replicated some. My idea is to start working at this as soon as I solve the thread subscription (I want to be sure who subscribe to a thread is the real owner of an email, and introduce also some limits to prevent spamming around). During the weekend I'll send the PR.

**Reply by danpros** [reply]
Currently, the date, tag, and slug form the .md filename, so it's safest to keep the slug as the ID. For example, if we add/remove a tag, update the date, etc., the filename will change. If the ID only uses a slug, then when we update a post slug, we just need to rename the comment JSON file with the new slug. I use a similar solution for the views counter. It is very rare that we change the slug after publication.

And why do we need to use `find_post()` to get the right post URL because in htmly we can set the slug prefix, we can use default prefix (`year/month/slug`) or `/any-string/slug`.

**Reply by rolandixor** [reply]
Am I correct in my initial deduction that the comments use Bootstrap for theming? I haven't tested them yet but was taking a brief look at the code.

**Reply by Emidio** [reply]
No, it shouldn't. It just show DIVs with classes, Bootstrap is demanded to template. Where did you catch Bootstrap code in comments?

**Reply by rolandixor** [reply]
> No, it shouldn't. It just show DIVs with classes, Bootstrap is demanded to template. Where did you catch Bootstrap code in comments?

I was looking at the pull request and so some bootstrap class names, that's where I was getting that impression. I'll look further into it though. Once I know what to target I can work on some CSS for the comments.
